// nodehandler-json/dist/index.js
'use strict';

var sdk = require('axway-flow-sdk');
var json = require('./json');

var nodes = sdk.init(module);

//
// JSON
//
nodes.add('json', {
	name: 'JSON',
	icon: 'json.svg',
	description: 'The JSON node contains methods for parsing JavaScript Object Notation (JSON) and converting values to JSON.',
	category: 'core'
}).method('stringify', {
	name: 'Stringify',
	description: 'The stringify method converts a JavaScript value to a JSON string.'
}).parameter('value', {
	description: 'The value to convert to a JSON string.'
}, true /* required */
).parameter('space', {
	description: 'A string or number object that\'s used to insert white space into the output JSON string for readability purposes. If this is a number, it indicates the number of space characters to use as white space; this number is capped at 10. If this is a string, its maximum length is 10;  the string is used as white space. If this parameter is not provided, no white space is used.',
	oneOf: [{
		type: 'string',
		minLength: 1,
		maxLength: 10
	}, {
		type: 'integer',
		minimum: 1,
		maximum: 10
	}]
}, false /* required */
).output('next', {
	name: 'Next',
	context: '$.json',
	schema: { type: 'string' }
}).action(json.stringify).method('parse', {
	name: 'Parse',
	description: 'The parse method parses a JSON string, constructing the JavaScript value or object described by the string.'
}).parameter('json', {
	description: 'The JSON string to parse.',
	type: 'string',
	format: 'multiline'
}, true /* required */
).output('next', {
	name: 'Next',
	context: '$.value',
	schema: {}
}).output('error', {
	name: 'Error',
	context: '$.error',
	description: 'This output is triggered if the input is not a valid JSON string. The output value is the error object.',
	schema: {}
}).action(json.parse);

exports = module.exports = nodes;
