// nodehandler-dot/dist/index.js
'use strict';

var sdk = require('axway-flow-sdk');
var dot = require('./dot');

var nodes = sdk.init(module);

// doT
nodes.add('doT', {
	name: 'Compose',
	icon: 'doT.png',
	description: 'Compose strings and objects using doT templates. See https://olado.github.io/doT.',
	category: 'core'
}).method('formatStr', {
	name: 'Format string',
	description: 'Compose a string by evaluating a template. See https://olado.github.io/doT.'
}).parameter('template', {
	description: 'The doT template.',
	type: 'string',
	format: 'multiline'
}, true /* required */
).parameter('data', {
	description: 'The data to evaluate the template with. Use $ to access the entire context.',
	type: 'object'
}, true /* required */
).output('next', {
	name: 'Next',
	context: '$.value',
	schema: {}
}).output('error', {
	name: 'Error',
	context: '$.error',
	description: 'This output is triggered if the evaluated template is not valid. The output value is error object.',
	schema: {}
}).action(dot.formatStr).method('formatObj', {
	name: 'Format object',
	description: 'Compose an object by evaluating a template. The evaluated template is JSON parsed and so must be a valid JSON encoded string. See https://olado.github.io/doT.'
}).parameter('template', {
	description: 'The doT template.',
	type: 'string',
	format: 'multiline'
}, true /* required */
).parameter('data', {
	description: 'The data to evaluate the template with. Use $ to access the entire context.',
	type: 'object'
}, true /* required */
).output('next', {
	name: 'Next',
	context: '$.value',
	schema: {}
}).output('error', {
	name: 'Error',
	context: '$.error',
	description: 'This output is triggered if the evaluated template is not a valid JSON string. The output value is the error object.',
	schema: {}
}).action(dot.formatObj);

exports = module.exports = nodes;
