// nodehandler-base64/dist/index.js
'use strict';

var sdk = require('axway-flow-sdk');
var base64Encode = require('./encode');
var bas64Decode = require('./decode');

exports = module.exports = sdk.init(module).add('base64', {
	name: 'Base64',
	icon: 'base64.svg',
	description: 'The Base64 node contains methods for base64 encoding and decoding data.',
	category: 'core'
})
// #encode
.method('encode', {
	name: 'Encode',
	description: 'Encodes base64 data'
}).parameter('data', {
	description: 'The data to encode',
	oneOf: [{ type: 'number' }, { type: 'boolean' }, { type: 'string' }, { type: 'object' }]
}, true).output('encoded', {
	name: 'Encoded',
	description: 'The base64 encoded data',
	context: '$.b64data',
	schema: {
		type: 'string'
	}
}).action(base64Encode)
// #decode
.method('decode', {
	name: 'Decode',
	description: 'Decodes base64 encoded data'
}).parameter('data', {
	description: 'The data to decode',
	type: 'string',
	format: 'multiline'
}, true).parameter('as', {
	description: 'Expected decoded data format',
	type: 'string',
	enum: ['buffer', 'object', 'string'],
	default: 'buffer'
}, false).output('decoded', {
	name: 'Decoded',
	description: 'The base64 decoded data',
	context: '$.decoded',
	schema: {
		type: 'string'
	}
}).output('error', {
	name: 'Error',
	context: '$.b64error',
	schema: {}
}).action(bas64Decode);
